src/main/java/htsjdk/samtools/SBIIndexWriter.java

```diff
@@ -0,0 +1,170 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package htsjdk.samtools;
+
+import htsjdk.samtools.util.BinaryCodec;
+import htsjdk.samtools.util.IOUtil;
+import htsjdk.samtools.util.RuntimeIOException;
+
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+/**
+ * Writes SBI files as understood by {@link SBIIndex}.
+ * <p>
+ * To use this class, first construct an instance from an output stream, and a desired granularity. Then for each
+ * record in the file being indexed, pass the virtual file offset of the record to the {@link #processRecord} method.
+ * The indexer will keep a count of the records passed in an index every <i>n</i>th record. When there are no records
+ * left call {@link #finish} to complete writing the index.
+ */
+public final class SBIIndexWriter {
+
+    // Default to a granularity level of 4096. This is generally sufficient
+    // for very large BAM files, relative to a maximum heap size in the
+    // gigabyte range.
+    public static final long DEFAULT_GRANULARITY = 4096;
+
+    static final byte[] EMPTY_MD5 = new byte[16];
+    static final byte[] EMPTY_UUID = new byte[16];
+
+    private final OutputStream out;
+    private final long granularity;
+    private final File tempOffsetsFile;
+    private final BinaryCodec tempOffsetsCodec;
+    private long prev = -1;
+    private long recordCount;
+    private long virtualOffsetCount;
+
+    /**
+     * Prepare to write an SBI index with the default granularity.
+     *
+     * @param out the stream to write the index to
+     */
+    public SBIIndexWriter(final OutputStream out) {
+        this(out, SBIIndexWriter.DEFAULT_GRANULARITY);
+    }
+
+    /**
+     * Prepare to write an SBI index.
+     *
+     * @param out         the stream to write the index to
+     * @param granularity write the offset of every <i>n</i>th record to the index
+     */
+    public SBIIndexWriter(final OutputStream out, final long granularity) {
+        this.out = out;
+        this.granularity = granularity;
+        try {
+            // Write the offsets to a temporary file, then write the entire file contents to the output stream at
+            // the end, once we know the number of offsets. This is more efficient than using a List<Long> for very
+            // large numbers of offsets (e.g. 10^8, which is possible for low granularity), since the list resizing
+            // operation is slow.
+            this.tempOffsetsFile = File.createTempFile("offsets-", ".headerless.sbi");
```

Comment:
It would be nice to use `Path` over `File` in new code where possible. E.g. use `Files.createTempFile()` here and `Files.newOutputStream()` below.

Reply:
Thanks for the suggestion @pshapiro4broad. I've made the change in the latest version of this PR.